sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvokers.java

```diff
@@ -310,78 +357,100 @@ protected StackManipulation invokeTargetMethod(MethodDescription instrumentedMet
           // Invoke the target method
           wrapWithUserCodeException(
-              MethodDelegationBinder.MethodInvoker.Simple.INSTANCE.invoke(targetMethod)),
-          MethodReturn.VOID);
+              new StackManipulation.Compound(
+                  MethodDelegationBinder.MethodInvoker.Simple.INSTANCE.invoke(targetMethod),
+                  // Return from the instrumented method
+                  TargetMethodAnnotationDrivenBinder.TerminationHandler.Returning.INSTANCE.resolve(
+                      Assigner.DEFAULT, instrumentedMethod, targetMethod))));
+    }
+
+    protected StackManipulation pushArgumentsOfInstrumentedMethod(MethodDescription targetMethod) {
+      return MethodVariableAccess.allArgumentsOf(targetMethod);
     }
   }
 
   /**
-   * Implements {@link DoFnInvoker#invokeSetup} or {@link DoFnInvoker#invokeTeardown} by delegating
-   * respectively to the {@link Setup} and {@link Teardown} methods.
+   * Passes parameters to the delegated method by downcasting each parameter of non-primitive type
+   * to its expected type.
    */
-  private static final class LifecycleMethodDelegation extends MethodDelegation {
-    private final DoFnSignature.LifecycleMethod signature;
-
-    private LifecycleMethodDelegation(@Nullable DoFnSignature.LifecycleMethod signature) {
-      this.signature = signature;
+  private static class DowncastingParametersMethodDelegation extends SimpleMethodDelegation {
+    protected DowncastingParametersMethodDelegation(Method method) {
+      super(method);
     }
 
     @Override
-    protected StackManipulation invokeTargetMethod(MethodDescription instrumentedMethod) {
-      MethodDescription targetMethod =
-          new MethodCall.MethodLocator.ForExplicitMethod(
-                  new MethodDescription.ForLoadedMethod(checkNotNull(signature).targetMethod()))
-              .resolve(instrumentedMethod);
-      return new StackManipulation.Compound(
-          wrapWithUserCodeException(
-              MethodDelegationBinder.MethodInvoker.Simple.INSTANCE.invoke(targetMethod)),
-          MethodReturn.VOID);
+    protected StackManipulation pushArgumentsOfInstrumentedMethod(MethodDescription targetMethod) {
+      List<StackManipulation> pushParameters = new ArrayList<>();
+      TypeList.Generic paramTypes = targetMethod.getParameters().asTypeList();
+      for (int i = 0; i < paramTypes.size(); i++) {
+        TypeDescription.Generic paramT = paramTypes.get(i);
+        pushParameters.add(MethodVariableAccess.of(paramT).loadOffset(i + 1));
+        if (!paramT.isPrimitive()) {
+          pushParameters.add(TypeCasting.to(paramT));
+        }
+      }
+      return new StackManipulation.Compound(pushParameters);
     }
   }
 
   /**
    * Wraps a given stack manipulation in a try catch block. Any exceptions thrown within the try are
    * wrapped with a {@link UserCodeException}.
+   *
+   * <p>Assumes that the stack manipulation contains a method return at the end.
```

Comment:
I'm guessing the assumption led to the simplification below? I'm a bit worried that it may not apply to all cases. Was there a reason we needed to remove the catchBlockEnd to make this PR work?